portal/portal-service-impl/impl/src/java/org/sakaiproject/portal/service/BullhornServiceImpl.java

```diff
@@ -0,0 +1,339 @@
+        = "INSERT INTO SOCIAL_ALERTS (FROM_USER,TO_USER,EVENT,REF,EVENT_DATE,URL) VALUES(?,?,?,?,?,?)";
+
+    private final static String ACADEMIC_INSERT_SQL
+        = "INSERT INTO ACADEMIC_ALERTS (FROM_USER,TO_USER,EVENT,REF,TITLE,SITE_ID,EVENT_DATE,URL) VALUES(?,?,?,?,?,?,?,?)";
+
+    private static final List<String> HANDLED_EVENTS = new ArrayList<String>();
+
+    private AnnouncementService announcementService;
+    private AssignmentService assignmentService;
+    private EntityManager entityManager;
+    private EventTrackingService eventTrackingService;
+    private MemoryService memoryService;
+    private ProfileConnectionsLogic profileConnectionsLogic;
+    private ProfileLinkLogic profileLinkLogic;
+    private ProfileWallLogic profileWallLogic;
+    private ProfileStatusLogic profileStatusLogic;
+    private UserDirectoryService userDirectoryService;
+    private SecurityService securityService;
+    private SessionManager sessionManager;
+    private ServerConfigurationService serverConfigurationService;
+    private SiteService siteService;
+    private SqlService sqlService;
+
+    public void init() {
+
+        if (serverConfigurationService.getBoolean("useBullhornAlerts", true)) {
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_WALL_ITEM_NEW);
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_WALL_ITEM_COMMENT_NEW);
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_STATUS_UPDATE);
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_FRIEND_REQUEST);
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_FRIEND_CONFIRM);
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_FRIEND_IGNORE);
+            HANDLED_EVENTS.add(ProfileConstants.EVENT_MESSAGE_SENT);
+            HANDLED_EVENTS.add(ContentHostingService.EVENT_RESOURCE_ADD);
+            HANDLED_EVENTS.add(AnnouncementService.SECURE_ANNC_ADD);
+            HANDLED_EVENTS.add(AssignmentConstants.EVENT_ADD_ASSIGNMENT);
+            HANDLED_EVENTS.add(AssignmentConstants.EVENT_GRADE_ASSIGNMENT_SUBMISSION);
+            eventTrackingService.addObserver(this);
+        }
+
+        if (serverConfigurationService.getBoolean("auto.ddl", true)) {
+            sqlService.ddl(this.getClass().getClassLoader(), "bullhorn_tables");
+        }
+    }
+
+    public void update(Observable o, final Object arg) {
+
+        if (arg instanceof Event) {
+            new Thread(() -> {
+                Event e = (Event) arg;
+                String event = e.getEvent();
+                String ref = e.getResource();
+                String[] pathParts = ref.split("/");
+                String from = e.getUserId();
+                long at = e.getEventTime().getTime();
+                try {
+                    if (HANDLED_EVENTS.contains(event)) {
+                        Cache countCache = memoryService.newCache("bullhorn_alert_count_cache");
+
+                        if (ProfileConstants.EVENT_WALL_ITEM_NEW.equals(event)) {
```

Comment:
A `switch` would be better than an `if/else if` (ideally ProfileConstants would be an enum and the we could get compile time checking that every case was catered for :()